cli/flox/src/config/mod.rs

```diff
@@ -189,132 +186,164 @@ pub enum ReadWriteError {
+    let cache_dir = flox_dirs.get_cache_home();
+    let data_dir = flox_dirs.get_data_home();
+    let state_dir = flox_dirs.get_state_home();
+
+    let config_dir = user_config_dir;
+
+    let mut builder = HierarchicalConfig::builder()
+        .set_default("default_substituter", "https://cache.floxdev.com/")?
+        .set_default("cache_dir", cache_dir.to_str().unwrap())?
+        .set_default("data_dir", data_dir.to_str().unwrap())?
+        .set_default("state_dir", state_dir.to_str().unwrap())?
+        // Config dir is added to the config for completeness;
+        // the config file cannot change the config dir.
+        .set_override("config_dir", config_dir.to_str().unwrap())?;
+    // Read System Config first
+    if let Some(system_config_dir) = system_config_dir {
+        builder = builder.add_source(
+            config::File::from(system_config_dir.join(FLOX_CONFIG_FILE))
+                .format(config::FileFormat::Toml)
+                .required(false),
+        );
+    };
+
+    // Read User Config
+    builder = builder.add_source(
+        config::File::from(config_dir.join(FLOX_CONFIG_FILE))
+            .format(config::FileFormat::Toml)
+            .required(false),
+    );
+
+    // Override via env variables
+    let builder = {
+        let mut flox_envs = env
+            .into_iter()
+            .filter_map(|(k, v)| k.strip_prefix("FLOX_").map(|k| (k.to_owned(), v)))
+            .collect::<Vec<_>>();
+        builder
+            .add_source(mk_environment(&mut flox_envs, "NIX"))
+            .add_source(mk_environment(&mut flox_envs, "GITHUB"))
+            .add_source(mk_environment(&mut flox_envs, "FEATURES"))
+            .add_source(
+                Environment::default()
+                    .source(Some(HashMap::from_iter(flox_envs)))
+                    .try_parsing(true),
+            )
+    };
+
+    let final_config = builder.build()?;
+    Ok(final_config)
+}
 
-        #[cfg(not(test))]
-        let reload = false;
+impl Config {
+    fn parse_with(
+        flox_dirs: &BaseDirectories,
+        user_config_dir: PathBuf,
+        system_config_dir: Option<PathBuf>,
+        env: impl IntoIterator<Item = (String, String)>,
+    ) -> Result<Config> {
+        let final_config =
+            raw_config_from_parts(flox_dirs, user_config_dir, system_config_dir, env)?;
 
-        let final_config = Self::raw_config(reload)?;
         let cli_config: Config = final_config
             .to_owned()
             .try_deserialize()
             .context("Could not parse config")?;
         Ok(cli_config)
     }
 
+    /// Creates a [Config] from the environment and config files
+    pub fn parse() -> Result<Config> {
+        let base_directories = BaseDirectories::with_prefix(FLOX_DIR_NAME)?;
+        Self::parse_with(
+            &base_directories,
+            locate_user_config_dir(&base_directories)?,
+            locate_system_config_dir()?,
+            env::vars(),
+        )
+    }
+
```

Comment:
question blocking: is there a good reason not to test this or have you tested it a lot manually? Seems like a pretty big change to have zero test changes

Reply:
added some tests for picking up the overrides. we still have some tests that use test_env which could now be replaced more explicitly.
Ill do another pass over the remaining tests when updating the `xdg` crate.